src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,241 @@
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+    } else if (rcThrottle > highThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    }
+
+    // if failsafe is active, and we get here, we are in landing mode, so descend at configured max rate
+    if (failsafeIsActive()) {
+        throttleAdjustmentFactor = -1.0f;
+    }
+
+    altHoldState.targetAltitudeDelta = throttleAdjustmentFactor * altholdConfig()->altHoldTargetAdjustRate * taskIntervalSeconds;
+    // if taskRate is 100Hz, default adjustRate of 100 adds/subtracts 1m every second, or 1cm per task run, at full stick position
+    altHoldState.targetAltitudeCm += altHoldState.targetAltitudeDelta;
+}
+
+void altHoldUpdate(void)
+{
+    // things that always need to happen in the background should go here
+    altHoldState.measuredAltitudeCm = getAltitude();
+    DEBUG_SET(DEBUG_ALTHOLD, 1, lrintf(altHoldState.measuredAltitudeCm));
+
+    altHoldProcessTransitions(); // initialise values, don't start without altitude data
+
+    // exit unless altHold is active
+    if (!altHoldState.isAltHoldActive) {
+        const uint8_t altitudeDataAvailable = isAltitudeAvailable();
+        DEBUG_SET(DEBUG_ALTHOLD, 0, altitudeDataAvailable);
+        return;
+    }
+
+    // if active, and we lose altitude readings (lose GPS 3D fix for example)
+    // if the last measured altitude was below target, we will climb indefinitelym, and vice versa
+    // we can set current altitude to target, making PIDs zero, and must force zero iTerm
+    // quad then gets the un-modified hover value and may climb or descend, but this should be slow
+    // user will havbe NO control user needs a warning (how? esp LOS) and should exit
+    if (!isAltitudeAvailable()) {
+        altHoldState.measuredAltitudeCm = altHoldState.targetAltitudeCm;
+        // force integral to zero or will accumulate
+        simplePid.integral = 0.0f;
+    }
+
+    // check if the user has changed the target altitude using sticks
+    if (altholdConfig()->altHoldTargetAdjustRate) {
+        altHoldUpdateTargetAltitude();
+    }
+
+    // use PIDs to return the throttle adjustment value, add it to the hover value, and constrain
+    const float throttleAdjustment = altitudePidCalculate();
+    float newThrottle = positionConfig()->hover_throttle + throttleAdjustment;
+    altHoldState.throttleOut = constrainf(newThrottle, altholdConfig()->altHoldThrottleMin, altholdConfig()->altHoldThrottleMax);
+
+    DEBUG_SET(DEBUG_ALTHOLD, 0, lrintf(altHoldState.targetAltitudeCm));
+    DEBUG_SET(DEBUG_ALTHOLD, 2, lrintf(throttleAdjustment));
+}
+
+void updateAltHoldState(timeUs_t currentTimeUs) {
+    altHoldUpdate();
+    UNUSED(currentTimeUs);
+}
+
+float altHoldGetThrottle(void) {
+    return scaleRangef(altHoldState.throttleOut, MAX(rxConfig()->mincheck, PWM_RANGE_MIN), PWM_RANGE_MAX, 0.0f, 1.0f);
```

Comment:
(it's time to move code to SI units and 0..1 scale values)

Reply:
However I do very much like that in this PR all the P, I, D and other adjustments are in 1000-2000 throttle units.  It makes life very simple when considering the logs, for instance.  However the code is more complex as a result.  I guess in a 0-1 situation the PID values are just 1000 times smaller.

I don't really understand the MAX for mincheck here - could anyone explain it?  For these purposes I don't know that it is relevant.  I just copied it from the GPS Rescue code.

Reply:
Result could be negative if throttleOut would be below mincheck. Guess it should be

```
return scaleRangef(MAX(altHoldState.throttleOut, rxConfig()->mincheck), PWM_RANGE_MIN, PWM_RANGE_MAX, 0.0f, 1.0f);
```

Reply:
Example:
```
With throttle 1050 and mincheck 1070

New Throttle -0.02 (this PR)
New Throttle 0.05 (no MAX check)
New Throttle 0.07 (above suggestion)

With throttle 1500 and mincheck 1070

New Throttle 0.46
New Throttle 0.50
New Throttle 0.50
```

Reply:
So IMO we can remove the MAX check

Reply:
Thinking about this...
`throttleOut` is already constrained between `alt_hold_throttle_min` and `alt_hold_throttle_max`, before it gets to the final scaler in `altHoldGetThrottle()`, so it can't be negative.
So we can never get a negative throttle to the scaler.
We could get a value below minCheck in some extreme settings.  
I'll limit the min and max setting ranges in CLI, let's see if anyone complains about the choices I made.
I guess the intent is to not go below minCheck in the value sent to the mixer.  I think that's because throttle control normally doesn't start until minCheck.  That would mean the output value we send would get scaled up a little bit compared to scaling only from PWM_RANGE_MIN.
@ledvinap may I ask for your thoughts about this? I think this seems OK at present.

Reply:
output is <= 0 for (throttleOut < mincheck) || (throttleOut < PWM_RANGE_MIN).
It may be correct - no output throttle below mincheck; PWM_RANGE_MIN is used when mincheck is missing.
from gps.c:
_We need to compensate for min_check since the throttle value set by gps rescue is based on the raw rcCommand value commanded by the pilot._

Ideally, throttleOut would be float between 0..1 and no conversion would be necessary here.

Also, `mincheck` is not used when 3d mode is active, it is ignored here.